### PR TITLE
Discovery endpoints

### PR DESCRIPTION
**Pull request by aphelionz**
Here's are some more endpoint functionality based on @horsburgh's request.

### CODE REVIEW

**Comment by aphelionz** [comment]
@pkdash have another look please

**Comment by pkdash** [comment]
@aphelionz I will try to finish reviewing today. Hope after that we will be able to release/publish 1.2.7.

File: docs/index.rst [hunk 1]
```diff
@@ -349,6 +349,22 @@ To set resource flags:
                   }
     >>> result = hs.resource('ID OF RESOURCE GOES HERE').flag(options)
 
+To discover resources via subject or bounding box:
```
Comment (pkdash) [inline docs/index.rst]: What about other search/query parameters like date range that is supported by the endpoint? Do we need show an example of using date range?

Comment (aphelionz) [inline docs/index.rst]: I can add those, yeah. They're the same as getResourceList

File: hs_restclient/endpoints/resources.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        return r
+
+
+class ResourceList(BaseEndpoint):
+    def __init__(self, hs, **kwargs):
+        super(ResourceList, self).__init__(hs)
+
+        """
+        Query the GET /hsapi/resource/ REST end point of the HydroShare server.
+
```
Comment (pkdash) [inline hs_restclient/endpoints/resources.py]: Should we be including other parameters as per the validator here (https://github.com/hydroshare/hydroshare/blob/develop/hs_core/views/serializers.py#L69)?

File: hs_restclient/endpoints/resources.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        :raises: HydroShareHTTPException to signal an HTTP error
+        :raises: HydroShareArgumentException for any invalid arguments
+
+        :return: A generator that can be used to fetch dict objects, each dict representing
+            the JSON object representation of the resource returned by the REST end point.  For example:
+
+        >>> for resource in hs.getResourceList():
+        >>>>    print resource
+         {u'bag_url': u'http://www.hydroshare.org/static/media/bags/e62a438bec384087b6c00ddcd1b6475a.zip',
```
Comment (pkdash) [inline hs_restclient/endpoints/resources.py]: Seems the return json data for resource is missing few attributes/keys (https://github.com/hydroshare/hydroshare/blob/develop/hs_core/views/serializers.py#L91)

**Comment by aphelionz** [comment]
@pkdash I believe I addressed your comments, have another look

File: hs_restclient/endpoints/resources.py [hunk 4]
```diff
@@ -181,6 +181,18 @@ def __init__(self, hs, **kwargs):
             resources created up to and including 2015-05-05)
         :param types: Filter results to particular HydroShare resource types.  Must be a sequence type
             (e.g. list, tuple, etc.), but not a string.
+        :param start: Filter results by start
+        :param count: Filter results by count
+        :param subject: Filter by comma separated list of subjects
+        :param metadata: Fillter by JSON metadata
```
Comment (pkdash) [inline hs_restclient/endpoints/resources.py]: typo

File: hs_restclient/endpoints/resources.py [hunk 5]
```diff
@@ -181,6 +181,18 @@ def __init__(self, hs, **kwargs):
             resources created up to and including 2015-05-05)
         :param types: Filter results to particular HydroShare resource types.  Must be a sequence type
             (e.g. list, tuple, etc.), but not a string.
+        :param start: Filter results by start
+        :param count: Filter results by count
+        :param subject: Filter by comma separated list of subjects
+        :param metadata: Fillter by JSON metadata
+        :param full_text_search: Filter by fill text search
```
Comment (pkdash) [inline hs_restclient/endpoints/resources.py]: fill -> full?

**Comment by pkdash** [comment]
@aphelionz Left couple minor comments. Otherwise +1.

**Pull request merged**